File: Data Structure & Algorithm [Old]/Summary List/Labuladong/Array/Two Points/Left & Right/_@5. Longest Palindromic Substring/5. Longest Palindromic Substring.go
// 5. Longest Palindromic Substring

// Given a string s, return the longest
// palindromic substring in s.

// Example 1:
// Input: s = "babad"
// Output: "bab"
// Explanation: "aba" is also a valid answer.

// Example 2:
// Input: s = "cbbd"
// Output: "bb"

package main

import "fmt"

// Find the longest palindrome in s with s[l] and s[r] as the center
func palindrome(s string, left int, right int) string {
	// Prevent index out of bounds
	for left >= 0 && right < len(s) && s[left] == s[right] {
		// Two pointers, expand to both sides
		left--
		right++
	}
	// Return the longest palindrome with s[l] and s[r] as the center
	return s[left+1 : right]
}

func longestPalindrome(s string) string {
	var res string
	for i := 0; i < len(s); i++ {
		// the longest palindromic substring centered at s[i]
		s1 := palindrome(s, i, i)
		// the longest palindromic substring centered at s[i] and s[i+1]
		s2 := palindrome(s, i, i+1)
		// res = longest(res, s1, s2)
		if len(res) < len(s1) {
			res = s1
		}
		if len(res) < len(s2) {
			res = s2
		}
	}
	return res
}

// test
func main() {
	res := longestPalindrome("babad")
	fmt.Println(res)
}

File: Data Structure & Algorithm/Archive/Data Structure/String/en/easy/5. Longest Palindromic Substring/5. Longest Palindromic Substring.ts
// 5. Longest Palindromic Substring

// Given a string s, return the longest palindromic substring in s.

// Example 1:
// Input: s = "babad"
// Output: "bab"
// Explanation: "aba" is also a valid answer.

// Example 2:
// Input: s = "cbbd"
// Output: "bb"

export function longestPalindrome(s: string): string {
  let maxLen: number = 0;
  let res: string = '';

  const isPalindromic = (s: string, left: number, right: number): boolean => {
    if (left >= right) {
      return true;
    }

    if (s.charAt(left) !== s.charAt(right)) {
      return false;
    }

    if (right - left <= 2) {
      return true;
    }

    return isPalindromic(s, left + 1, right - 1);
  };

  for (let left: number = 0; left < s.length; left++) {
    for (let right: number = left; right < s.length; right++) {
      if (s.charAt(left) === s.charAt(right) && isPalindromic(s, left + 1, right - 1)) {
        if (maxLen < right - left + 1) {
          maxLen = right - left + 1;
          res = s.substring(left, right + 1);
        }
      }
    }
  }

  return res;
}

// test
const res = longestPalindrome('babad');
console.log(res);
